src/printer.js

```diff
@@ -1995,93 +1992,140 @@ function printMemberLookup(path, print) {
 //   .filter(x => x > 10)
 //   .some(x => x % 2)
 //
-// where there is a "chain" of function calls on the result of
-// previous expressions. We want to format them like above so they
-// are consistently aligned.
-//
-// The way we do is by eagerly traversing the AST tree and
-// re-shape it into a list of calls on member expressions. This
-// lets us implement a heuristic easily for when we want to format
-// it: if there are more than 1 calls on a member expression that
-// pass in a function, we treat it like the above.
+// The way it is structured in the AST is via a nested sequence of
+// MemberExpression and CallExpression. We need to traverse the AST
+// and make groups out of it to print it in the desired way.
 function printMemberChain(node, options, print) {
-  const nodes = [];
-  let leftmost = node;
-  let leftmostParent = null;
-  // Traverse down and gather up all of the calls on member
-  // expressions. This flattens it out into a list that we can
-  // easily analyze.
-  while (
-    leftmost.type === "CallExpression" &&
-      leftmost.callee.type === "MemberExpression"
-  ) {
-    nodes.push({ member: leftmost.callee, call: leftmost });
-    leftmostParent = leftmost;
-    leftmost = leftmost.callee.object;
-  }
-  nodes.reverse();
-
-  // There are two kinds of formats we want to specialize: first,
-  // if there are multiple calls on lookups we want to group them
-  // together so they will all break at the same time. Second,
-  // it's a chain if there 2 or more calls pass in functions and
-  // we want to forcibly break all of the lookups onto new lines
-  // and indent them.
-  function argIsFunction(call) {
-    if (call.arguments.length > 0) {
-      const type = call.arguments[0].type;
-      return typeIsFunction(type);
+  // The first phase is to linearize the AST by traversing it down.
+  //
+  //   a().b()
+  // has the following AST structure:
+  //   CallExpression(MemberExpression(CallExpression(Identifier)))
+  // and we transform it into
+  //   [Identifier, CallExpression, MemberExpression, CallExpression]
+  const printedNodes = [];
+  while (true) {
+    if (node.type === "CallExpression") {
+      printedNodes.unshift({
+        type: node.type,
+        printed: printArgumentsList(FastPath.from(node), options, print)
```

Comment:
You can't just create a path from a node. See the previous code. You have to create it from the parent and then call the node.

This is because otherwise lots of code that calls `getParentNode` will return null because the node doesn't have a parent in the path.